linkerd/access-log/src/tower.rs

```diff
@@ -0,0 +1,169 @@
+/// A tower layer that associates a tokio-tracing Span with each request
+#[derive(Clone)]
+pub struct AccessLogLayer {}
+
+#[derive(Clone)]
+pub struct AccessLogContext<Svc> {
+    inner: Svc,
+}
+
+struct ResponseFutureInner {
+    span: Span,
+    start: Instant,
+    processing: Duration,
+}
+
+#[pin_project]
+pub struct AccessLogFuture<F> {
+    data: Option<ResponseFutureInner>,
+
+    #[pin]
+    inner: F,
+}
+
+impl<Svc> tower::layer::Layer<Svc> for AccessLogLayer {
+    type Service = AccessLogContext<Svc>;
+
+    fn layer(&self, inner: Svc) -> Self::Service {
+        Self::Service { inner }
+    }
+}
+
+impl<Svc, B1, B2> tower::Service<http::Request<B1>> for AccessLogContext<Svc>
+where
+    Svc: tower::Service<http::Request<B1>, Response = http::Response<B2>>,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = AccessLogFuture<Svc::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Svc::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: http::Request<B1>) -> Self::Future {
+        let span: Span = span!(target: "access_log", Level::TRACE, "http",
+            timestamp=field::Empty, processing_ns=field::Empty, total_ns=field::Empty,
+            method=field::Empty, uri=field::Empty, version=field::Empty, user_agent=field::Empty,
+            host=field::Empty, trace_id=field::Empty, status=field::Empty,
+            request_bytes=field::Empty, response_bytes=field::Empty);
+
+        if span.is_disabled() {
+            return AccessLogFuture {
+                data: None,
+                inner: self.inner.call(request),
+            };
+        }
+
+        // Delay formatting to avoid an intermediate `String`
+        let delayed_format = chrono::Utc::now().format_with_items(
+            [chrono::format::Item::Fixed(chrono::format::Fixed::RFC3339)].iter(),
+        );
+
+        span.record("timestamp", &field::display(&delayed_format));
+        span.record("method", &request.method().as_str());
+        span.record("uri", &field::display(&request.uri()));
+        span.record("version", &field::debug(&request.version()));
+
+        request
+            .headers()
+            .get("Host")
+            .and_then(|x| x.to_str().ok())
+            .map(|x| span.record("host", &x));
+
+        request
+            .headers()
+            .get("User-Agent")
+            .and_then(|x| x.to_str().ok())
+            .map(|x| span.record("user_agent", &x));
+
+        request
+            .headers()
+            .get("Content-Length")
+            .and_then(|x| x.to_str().ok())
+            .map(|x| span.record("request_bytes", &x));
+
+        request
+            .headers()
+            .get("x-b3-traceid")
+            .or_else(|| request.headers().get("X-Request-ID"))
+            .or_else(|| request.headers().get("X-Amzn-Trace-Id"))
+            .and_then(|x| x.to_str().ok())
+            .map(|x| span.record("trace_id", &x));
```

Comment:
This could probably be much simpler without sacrificing performance. My guess is that we're using `field::Empty` here to avoid evaluating the fields if the span is disabled. But, if the span is disabled, the `tracing::span!` macro won't evaluate any field expressions, so i think we could pass most of these (or at least, the simpler ones) as fields in the macro.

Also, in the case where the span _is_ enabled, it's actually more efficient to record as many fields in the `span!` macro as possible, since they will be batched and passed to the `tracing` subscriber in one function call. When we call the `record` method a bunch of times, we pass each individual field on its own.

Reply:
I wasn't aware that the macro wouldn't evaluate fields if the span is disabled - will fix